CONTRIBUTE.md

```diff
@@ -0,0 +1,126 @@
+  following order:
+  1. `KALE_PIP_INDEX_URLS` – comma separated list, highest priority.
+  2. `KALE_DEV_MODE` + optional `KALE_DEVPI_SIMPLE_URL` – enables your local devpi.
+  3. Default fallback – `https://pypi.org/simple`.
+
+Because compiled workflows freeze both the Kale version and the index list,
+updating your local package copy requires re‑publishing the wheel/sdist that the
+pipeline will install. That is why a disposable package index (devpi) is useful –
+you can iterate without uploading experimental builds to the public PyPI.
+
+## 5. Working with the local devpi index
+
+The `backend/scripts/` directory provides helpers for spinning up a local devpi
+registry and publishing the freshly built Kale package:
+
+| Script | Purpose | Typical usage |
+| ------ | ------- | ------------- |
+| `devpi-up.sh` | Start (or reuse) a devpi-server instance on `HOST:PORT` and create a volatile `root/<index>` derived from the public PyPI mirror. Prints the environment variables needed for pip. | `./scripts/devpi-up.sh` |
+| `devpi-publish.sh` | Build Kale (`python -m build`), infer the newest artifact version from `dist/`, remove any existing release with the same version from the target devpi index, then upload the artifacts. | `./scripts/devpi-publish.sh` |
+| `devpi-main.sh` | Convenience wrapper that calls the two scripts above and exports `KALE_DEV_MODE=1` + `KALE_DEVPI_SIMPLE_URL` in the current shell. | `source ./scripts/devpi-main.sh` |
+
+Recommended loop when iterating on compiled pipelines:
+
+```bash
+source .venv/bin/activate
+cd backend
+
+# Start devpi and export KALE_DEV_MODE / KALE_DEVPI_SIMPLE_URL / pip overrides.
+source scripts/devpi-main.sh
+
+# Make your code changes…
+# Rebuild and publish to the local index (removes the previous dev build).
+./scripts/devpi-publish.sh
+
+# Re-run notebook compilation so the generated component pulls from devpi.
+kale --help  # or your usual CLI invocation pointing at the notebook
+# For example you can run this quick test to validate the the generated DSL points
+# for the correct Kale version
+# kale --nb ./examples/base/candies_sharing.ipynb --dev
+```
+
+> Note: the scripts rely on `devpi-server`, `devpi-client`, and `python -m build`.
+> Install them once in your virtualenv: `pip install devpi-server devpi-client build`.
+
+## 6. Running tests
+
+```bash
+cd backend
+python -m pytest backend/kale/tests/unit_tests
+```
+
+Integration tests under `backend/kale/tests/e2e/` perform notebook → DSL
+compilation and compare against golden files in `backend/kale/tests/assets/`.
+If you alter template output, update the generated files accordingly.
+
+## 7. Formatting and linting
+
+- Python formatting is enforced via the existing code style (PEP 8). Many generated
+  files run through `autopep8`.
+- The labextension has independent tooling (`jlpm lint`, `jlpm prettier`). Run
+  those only if you modify the frontend package.
+
+## 8. Typical development checklist
```

Comment:
I tried to follow these steps quite a few times and I noticed that the pipeline generated by Kale will still refer to localhost(127.0.0.1) devpi. But it fails due the fact that KFP is adding the pip index-url automatically to trusted host (see this [issue](https://github.com/kubeflow/pipelines/issues/11155)). 

In any case, if we fix that, we would still have to figure out how to fix the problem with an accessible devpi for KFP - in this case I think that the best solution would be running devpi on a pod in kubeflow namespace. (great thing that you made INDEX_URL a parameter!)

Does this make sense?